src/annotator/anchoring/pdf.js

```diff
@@ -1,44 +1,70 @@
 /* global PDFViewerApplication */
 
-import seek from 'dom-seek';
-
 import RenderingStates from '../pdfjs-rendering-states';
 
-import { BrowserRange } from './range';
-import { toRange as textPositionToRange } from './text-position';
-import { TextPositionAnchor, TextQuoteAnchor } from './types';
+import { TextPosition, TextRange } from './text-range';
+import { TextQuoteAnchor } from './types';
 
 /**
  * @typedef {import('../../types/api').TextPositionSelector} TextPositionSelector
  * @typedef {import('../../types/api').TextQuoteSelector} TextQuoteSelector
+ * @typedef {import('../../types/api').Selector} Selector
  *
  * @typedef {import('../../types/pdfjs').PDFPageView} PDFPageView
  * @typedef {import('../../types/pdfjs').PDFViewer} PDFViewer
  */
 
+/**
+ * @typedef PdfTextRange
+ * @prop {number} pageIndex
+ * @prop {Object} anchor
+ * @prop {number} anchor.start
+ * @prop {number} anchor.end
+ */
+
 // Caches for performance.
 
 /**
- * Map of page index to page text content as a `Promise<string>`.
+ * Map of page index to page text content.
+ *
+ * @type {Record<number,Promise<string>>}
  */
 let pageTextCache = {};
 
 /**
- * 2D map from `[quote][position]` to `{pageIndex, anchor}` intended to optimize
- * re-anchoring of a pair of quote and position selectors if the position
- * selector fails to anchor on its own.
+ * A cache that associates the quote and position from an annotation's
+ * selectors to the resolved location in the PDF.
+ *
+ * This helps with re-anchoring of a pair of quote and position selectors if
+ * the position selector fails to anchor.
+ *
+ * @type {Record<string, Record<number, PdfTextRange>>}
  */
 let quotePositionCache = {};
 
+/**
+ * Return offset of `node` among its siblings
+ *
+ * @param {Node} node
+ */
 function getSiblingIndex(node) {
-  return Array.from(node.parentNode.childNodes).indexOf(node);
+  let index = 0;
+  while (node.previousSibling) {
+    ++index;
+    node = node.previousSibling;
+  }
+  return index;
 }
 
+/**
+ * Return the text layer element of the PDF page containing `node`.
+ *
+ * @param {Node|Element} node
+ * @return {Element|null}
+ */
 function getNodeTextLayer(node) {
-  while (!node.classList || !node.classList.contains('page')) {
-    node = node.parentNode;
-  }
-  return node.getElementsByClassName('textLayer')[0];
+  const el = 'closest' in node ? node : node.parentElement;
+  return el?.closest('.textLayer') ?? null;
```

Comment:
This change ensures that `getNodeTextLayer` only returns an `Element` if the passed `node` is contained _within_ the text layer. The rest of the code in `describe` after `getNodeTextLayer` didn't handle all the possible cases when the range started or ended outside the text layer, leading to issues like https://github.com/hypothesis/client/issues/1464.

Reply:
Can you confirm that changes here and in `getSiblingIndex` are orthogonal to the main type-related changes in this diff?